python/cuml/ts/batched_lbfgs.py

```diff
@@ -0,0 +1,122 @@
+import scipy.optimize as optimize
+import numpy as np
+from IPython.core.debugger import set_trace
+from scipy.optimize import _lbfgsb
+from .batched_kalman import pynvtx_range_push, pynvtx_range_pop
+
+from collections import deque
+
+def _fd_fprime(x, f, h):
+    g = np.zeros(len(x))
+    for i in range(len(x)):
+        xph = np.copy(x)
+        xmh = np.copy(x)
+        xph[i] += h
+        xmh[i] -= h
+        fph = f(xph)
+        fmh = f(xmh)
+        g[i] = (fph - fmh)/(2*h)
+
+    return g
+
+def batched_fmin_lbfgs_b(func, x0, num_batches, fprime=None, args=(),
```

Comment:
How much time is spent using Scipy's CPU batched lbfgs? If our initial version uses Scipy's still when merged, please create an issue to move this to GPU batched lbfgs

Reply:
This is actually a pretty tricky point. SciPy's L-BFGS implementation is quite robust, and my own attempts to replicate a simpler version of it were unsuccessful. We could port the implementation, but unfortunately the F77 internals of SciPy's L-BFGS implementation are quite a project to understand and rewrite. Generally the time-spent inside of L-BFGS is negligible, however the iteration count to convergence directly impacts performance.

Do we not have SciPy as a dependency? If not, we'll have to see what to do here. I guess the license of the F77 code would allow us to use it directly, but I don't know if we want to have a Fortran compiler as an additional dependency.

Reply:
I don’t think adding scipy as a dependency shouldn’t be much of an issue, license or technical wise. I guess what I worry the most is not the quality of scipys implementation, but the cost of having to move memory from device to host to device, which o guess might be more relevant for an eventual multi gpu version even more.